### PR TITLE
Windows Notifications

### PR DESCRIPTION
**Pull request by felixrieseberg**
This is now ready for review.

### CODE REVIEW

**Comment by felixrieseberg** [comment]
My :european_castle: for anyone who can tell me how we can link `WindowsCreateStringReference` (and its friends).

```
FAILED: C:\Python27\python.exe gyp-win-tool link-with-manifests environment.x86 True electron.exe "C:\Python27\python.exe gyp-win-tool link-wrapper environment.x86 False link.exe /nologo /OUT:electron.exe @electron.exe.rsp" 1 mt.exe rc.exe "obj\electron.electron.exe.intermediate.manifest" obj\electron.electron.exe.generated.manifest ..\..\atom\browser\resources\win\atom.manifest
brightray.lib(brightray.notification_presenter_win.obj) : error LNK2019: unresolved external symbol __imp__RoGetActivationFactory@12 referenced in function "long __cdecl Windows::Foundation::GetActivationFactory<struct ABI::Windows::UI::Notifications::IToastNotificationFactory>(struct HSTRING__ *,struct ABI::Windows::UI::Notifications::IToastNotificationFactory * *)" (??$GetActivationFactory@UIToastNotificationFactory@Notifications@UI@Windows@ABI@@@Foundation@Windows@@YAJPAUHSTRING__@@PAPAUIToastNotificationFactory@Notifications@UI@1ABI@@@Z)
brightray.lib(brightray.notification_presenter_win.obj) : error LNK2019: unresolved external symbol _WindowsCreateStringReference@16 referenced in function "public: __thiscall StringReferenceWrapper::StringReferenceWrapper<4>(wchar_t const (&)[4])" (??$?0$03@StringReferenceWrapper@@QAE@AAY03$$CB_W@Z)
brightray.lib(brightray.notification_presenter_win.obj) : error LNK2019: unresolved external symbol _WindowsDeleteString@4 referenced in function "public: __thiscall StringReferenceWrapper::~StringReferenceWrapper(void)" (??1StringReferenceWrapper@@QAE@XZ)
electron.exe : fatal error LNK1120: 3 unresolved externals
Traceback (most recent call last):
  File "gyp-win-tool", line 316, in <module>
    sys.exit(main(sys.argv[1:]))
  File "gyp-win-tool", line 29, in main
    exit_code = executor.Dispatch(args)
  File "gyp-win-tool", line 71, in Dispatch
    return getattr(self, method)(*args[1:])
  File "gyp-win-tool", line 169, in ExecLinkWithManifests
    subprocess.check_call(ldcmd + add_to_ld)
  File "C:\Python27\lib\subprocess.py", line 511, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command 'C:\Python27\python.exe gyp-win-tool link-wrapper environment.x86 False link.exe /nologo /OUT:electron.exe @electron.exe.rsp electron.exe.manifest.res' returned non-zero exit status 1120
ninja: build stopped: subcommand failed.
```

**Comment by paulcbetts** [comment]
When in doubt, abuse FINDSTR:

```
C:\Program Files (x86)\Windows Kits\10> findstr /M /S WindowsCreateStringReference  *
App Certification Kit\aeinv.dll
App Certification Kit\aitstatic.exe
App Certification Kit\AllowList81.xml
App Certification Kit\appxman.dll
App Certification Kit\microsoft.windows.softwarelogo.metadatavalidation.tests.dll
App Certification Kit\SupportedAPIs-arm.xml
App Certification Kit\SupportedAPIs-x64.xml
App Certification Kit\SupportedAPIs-x86.xml
App Certification Kit\SupportedAPIsXML-6.2.0.xml
App Certification Kit\SupportedAPIsXML-6.3.0.xml
bin\arm\AppPerfAnalyzer\appperfanalyzer_js.dll
bin\arm\DeployUtil.exe
bin\arm\XamlDiagnostics\xamldiagnostics.dll
bin\arm64\AppPerfAnalyzer\appperfanalyzer_js.dll
bin\arm64\DeployUtil.exe
bin\arm64\XamlDiagnostics\xamldiagnostics.dll
bin\x64\AppPerfAnalyzer\appperfanalyzer_js.dll
bin\x64\DeployUtil.exe
bin\x64\XamlDiagnostics\xamldiagnostics.dll
bin\x86\AppPerfAnalyzer\appperfanalyzer_js.dll
bin\x86\DeployUtil.exe
bin\x86\XamlDiagnostics\xamldiagnostics.dll
Debuggers\x64\plmdebug.exe
Debuggers\x86\plmdebug.exe
Extension SDKs\WindowsMobile\10.0.10240.0\Lib\um\arm\onecoreuap.lib
Extension SDKs\WindowsMobile\10.0.10240.0\Lib\um\x86\onecoreuap.lib
Include\10.0.10240.0\winrt\winstring.h
Include\10.0.10240.0\winrt\wrl\wrappers\corewrappers.h
Lib\10.0.10240.0\ucrt\arm\libucrt.lib
Lib\10.0.10240.0\ucrt\arm\libucrtd.lib
Lib\10.0.10240.0\ucrt\arm64\libucrt.lib
Lib\10.0.10240.0\ucrt\arm64\libucrtd.lib
Lib\10.0.10240.0\ucrt\x64\libucrt.lib
Lib\10.0.10240.0\ucrt\x64\libucrtd.lib
Lib\10.0.10240.0\ucrt\x86\libucrt.lib
Lib\10.0.10240.0\ucrt\x86\libucrtd.lib
Lib\10.0.10240.0\um\arm\mincore.lib
Lib\10.0.10240.0\um\arm\OneCore.Lib
Lib\10.0.10240.0\um\arm\OneCoreUAP.Lib
Lib\10.0.10240.0\um\arm\runtimeobject.lib
Lib\10.0.10240.0\um\arm\WindowsApp.lib
Lib\10.0.10240.0\um\arm64\mincore.lib
Lib\10.0.10240.0\um\arm64\OneCore.Lib
Lib\10.0.10240.0\um\arm64\OneCoreUAP.Lib
Lib\10.0.10240.0\um\arm64\runtimeobject.lib
Lib\10.0.10240.0\um\arm64\WindowsApp.lib
Lib\10.0.10240.0\um\x64\mincore.lib
Lib\10.0.10240.0\um\x64\OneCore.Lib
Lib\10.0.10240.0\um\x64\OneCoreUAP.Lib
Lib\10.0.10240.0\um\x64\runtimeobject.lib
Lib\10.0.10240.0\um\x64\WindowsApp.lib
Lib\10.0.10240.0\um\x86\mincore.lib
Lib\10.0.10240.0\um\x86\OneCore.Lib
Lib\10.0.10240.0\um\x86\OneCoreUAP.Lib
Lib\10.0.10240.0\um\x86\runtimeobject.lib
Lib\10.0.10240.0\um\x86\WindowsApp.lib
Testing\Runtimes\TAEF\MinTe\Wex.Common.dll
Testing\Runtimes\TAEF\Te.WinRT.dll
Testing\Runtimes\TAEF\Wex.Common.dll
Testing\Runtimes\TAEF\x64\MinTe\Wex.Common.dll
Testing\Runtimes\TAEF\x64\Te.WinRT.dll
Testing\Runtimes\TAEF\x64\Wex.Common.dll

C:\Program Files (x86)\Windows Kits\10> & 'C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\bin\amd64\dumpbin.exe' /EXPORTS Lib\10.0.10240.0\um\x64\windowsapp.lib | grep 'WindowsCreateString'
                  WindowsCreateString
                  WindowsCreateStringReference
```

**Comment by jpoon** [comment]
@paulcbetts How does one go about linking the lib that you referenced?

Felix and I added the lib that you had found to [brightray.gyp](https://github.com/atom/brightray/blob/windows-notifications-lib/brightray.gyp) (line 196):
>            'VCLinkerTool': {
                    'AdditionalDependencies': [
                      'advapi32.lib',
                      'dbghelp.lib',
                      'delayimp.lib',
                      'dwmapi.lib',
                      'gdi32.lib',
                      'netapi32.lib',
                      'oleacc.lib',
                      'powrprof.lib',
                      'user32.lib',
                      'usp10.lib',
                      'version.lib',
                      'winspool.lib',
       ---->          'windowsapp.lib',
                    ],

But still the same linking errors. Any way to get better console out from gyp/ninja?

**Comment by paulcbetts** [comment]
@jpoon Good question - maybe we're linking against an older SDK?

**Comment by zcbenz** [comment]
For the new added libraries, you probably need to add them as global link settings:

```python
        ['OS=="win"', {
          'msvs_settings': {
            'VCLinkerTool': {
              'AdditionalDependencies': [
                'runtimeobject.lib',
              ],
            },
          },
          'conditions':  [
            ['libchromiumcontent_component', {
```

Otherwise they will only be linked in Release build.

**Comment by felixrieseberg** [comment]
Quick update for everyone following along: It's building, just not working like we want it to. But we're getting closer.

**Comment by jpoon** [comment]
We have native notifications! :fireworks:  This is what I see on Windows 10:

untitled

Still some stuff left to do:
- [ ] Set the body/text of the toast
- [ ] Handle callbacks
- [ ] Verify behaviour on Win8, Win7

**Comment by felixrieseberg** [comment]
Rebased and added some stuff back - we now have basic notifications in.

**Comment by felixrieseberg** [comment]
notifications

**Comment by felixrieseberg** [comment]
@paulcbetts @zcbenz @kevinsawicki :stars: :fireworks: We *think* that we're done here. I know there's a big merge button right on this page, but I'd never merge it without y'all taking a look and giving me expressed consent.

If you want to try this puppy out, run the following:

```js
var windowsNotification = new Notification("Test Title", {
 	body: "Hi, I'm an example body. How are you?",
 	icon: "file:///C:/Path/To/Your/Image.png"
});

windowsNotification.onshow  = function () { console.log("Notification shown") };
windowsNotification.onclick = function () { console.log("Notification clicked") };
windowsNotification.onclose = function () { console.log("Notification dismissed") };
```

The obvious next step is some documentation, which we'll also gladly cover.

**Comment by paulcbetts** [comment]
Need to add a runtime check to see if you're on Windows 7 or else you'll segfault

File: browser/win/windows_toast_notification.cc [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+using namespace ABI::Windows::Data::Xml::Dom;
+
+#define BREAK_IF_BAD(hr)    if(!SUCCEEDED(hr)) break;
+
+namespace WinToasts {
+
+WindowsToastNotification::WindowsToastNotification(const char* appName, content::DesktopNotificationDelegate* delegate)
+{
+    // Init using WRL
+    Windows::Foundation::Initialize(RO_INIT_MULTITHREADED);
```
Comment (paulcbetts) [inline browser/win/windows_toast_notification.cc]: How many times are we calling this? Should this be statically initialized?

File: browser/win/notification_presenter_win.cc [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+void NotificationPresenterWin::ShowNotification(
+  const content::PlatformNotificationData& data,
+  const SkBitmap& icon,
+  scoped_ptr<content::DesktopNotificationDelegate> delegate_ptr,
+  base::Closure* cancel_callback) {
+  
+  std::wstring title = data.title;
+  std::wstring body = data.body;
+  std::string iconPath = data.icon.spec();
```
Comment (paulcbetts) [inline browser/win/notification_presenter_win.cc]: Is every icon guaranteed to have a path? NativeImages can be created from non-file sources, no?

Comment (felixrieseberg) [inline browser/win/notification_presenter_win.cc]: According to the spec, [the `icon` should only contain a URL](https://notifications.spec.whatwg.org/#dom-notification-icon). One could easily pass a non-file one, but in that case, Windows won't display it.

I briefly thought about some silly business involving downloading the image, saving it to a temp folder, and then passing the file string - but that's certainly a `next` feature, given how much padding we'd have to add to make that work reliably.

Comment (felixrieseberg) [inline browser/win/notification_presenter_win.cc]: Oh, and just to add to that: data.icon is a `Gurl` ;-)

Comment (paulcbetts) [inline browser/win/notification_presenter_win.cc]: :+1:

File: browser/win/windows_toast_notification.cc (hunk 1 above)
Comment (jpoon) [inline browser/win/windows_toast_notification.cc]: Good call. Statically initialized this to only run this once.

**Comment by jpoon** [comment]
Thanks for the comments @paulcbetts. Pushed an update to address the concerns you brought up.

File: browser/win/notification_presenter_win.cc [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+NotificationPresenterWin::~NotificationPresenterWin() {
+}
+
+void NotificationPresenterWin::ShowNotification(
+  const content::PlatformNotificationData& data,
+  const SkBitmap& icon,
+  scoped_ptr<content::DesktopNotificationDelegate> delegate_ptr,
+  base::Closure* cancel_callback) {
```
Comment (zcbenz) [inline browser/win/notification_presenter_win.cc]: The `cancel_callback` should be assigned to a function that cancels the notification, we can bind it to `RemoveNotification`:

```c++
*cancel_callback = base::Bind(&NotificationPresenterWin::RemoveNotification, base::Unretained(this));
```

Comment (jpoon) [inline browser/win/notification_presenter_win.cc]: :+1: good point. updated PR with change.

**Comment by felixrieseberg** [comment]
Nice! Thanks everyone (especially @paulcbetts and @jpoon). Code review fixes are in, commits are squashed. Ready for merge?

**Comment by zcbenz** [comment]
:+1:

**Comment by paulcbetts** [comment]
@felixrieseberg This is still going to crash *hard* on Windows 7 if anyone uses it. Can you submit another PR to make sure that we no-op this?

**Comment by miniak** [comment]
You can use WRL::RuntimeClass:
`class ToastEventHandler : public RuntimeClass<WRL::RuntimeClassFlags<WRL::ClassicCom>, DesktopToastActivatedEventHandler, DesktopToastDismissedEventHandler>`
go the the implementation of `AddRef`, `Release` and `QueryInterface` generated automatically.

**Comment by miniak** [comment]
https://github.com/atom/brightray/pull/166

**Comment by miniak** [comment]
I also don't think it's a good idea to keep a raw pointer to a COM object, which is probably gets AddRef-ed by `IToastNotification` when you pass the `ToastEventHandler` instance to the `add_Activated` and `add_Dismissed` methods. It should be stored in a WRL::ComPtr instead.

**Comment by miniak** [comment]
It looks like the implementation is based on https://code.msdn.microsoft.com/windowsdesktop/sending-toast-notifications-71e230a2, in that case I also wonder why you didn't keep `StringReferenceWrapper`. Why do you keep calling `WindowsDeleteString` manually every time? That's really stupid.

**Comment by miniak** [comment]
Actually, according to MSDN:
> You don't need to call the WindowsDeleteString function to de-allocate a fast-pass HSTRINGcreated by the WindowsCreateStringReference function.

**Comment by miniak** [comment]
To make sure that it does not break Windows 7, you can do something like this:

```
typedef decltype(&RoGetActivationFactory) RoGetActivationFactoryPtr;
typedef decltype(&RoInitialize) RoInitializePtr;
typedef decltype(&RoUninitialize) RoUninitializePtr;
typedef decltype(&WindowsCreateStringReference) WindowsCreateStringReferencePtr;

HRESULT WINAPI RoGetActivationFactory(HSTRING activatableClassId, REFIID iid, void ** factory)
{
    static auto hLib = LoadLibrary(L"api-ms-win-core-winrt-l1-1-0.dll");
    static auto func = reinterpret_cast<RoGetActivationFactoryPtr>(GetProcAddress(hLib, "RoGetActivationFactory"));

    return func ? func(activatableClassId, iid, factory) : E_FAIL;
}

HRESULT WINAPI RoInitialize(RO_INIT_TYPE initType)
{
    static auto hLib = LoadLibrary(L"api-ms-win-core-winrt-l1-1-0.dll");
    static auto func = reinterpret_cast<RoInitializePtr>(GetProcAddress(hLib, "RoInitialize"));

    return func ? func(initType) : E_FAIL;
}

void WINAPI RoUninitialize()
{
    static auto hLib = LoadLibrary(L"api-ms-win-core-winrt-l1-1-0.dll");
    static auto func = reinterpret_cast<RoUninitializePtr>(GetProcAddress(hLib, "RoUninitialize"));

    return func ? func() : (void)0;
}

STDAPI WindowsCreateStringReference(PCWSTR sourceString, UINT32 length, HSTRING_HEADER * hstringHeader, HSTRING * string)
{
    static auto hLib = LoadLibrary(L"api-ms-win-core-winrt-string-l1-1-0.dll");
    static auto func = reinterpret_cast<WindowsCreateStringReferencePtr>(GetProcAddress(hLib, "WindowsCreateStringReference"));

    return func ? func(sourceString, length, hstringHeader, string) : E_FAIL;
}
```

**Comment by miniak** [comment]
To fix inconsistent DLL linkage errors you have to `#define _ROAPI_` before including wrl.h

```
roapi.h:

...
#ifdef _ROAPI_
#define ROAPI
#else
#define ROAPI DECLSPEC_IMPORT
#endif // _ROAPI_
...
```

**Comment by miniak** [comment]
Oh, I see. You are using `WindowsCreateString` instead of `WindowsCreateStringReference`, which would have been more efficient.

**Pull request merged**